### DISCUSSION TITLE
Multiple account support for EX

### DISCUSSION THREAD

**Discussion by Thatoo** [element-hq/element-meta#1832] [Ideas]
Hopefully, both ElementX on iOS and Android will be the first Element clients to support Multiple accounts feature \(and the web client about that time too?\)

Just for the record, it is one of the oldest and a wanted feature in Element-ios,    https://github.com/vector-im/element-ios/issues/315, Element-android, https://github.com/vector-im/element-android/issues/26 \[and also Element-web but that's out of topic... https://github.com/vector-im/element-web/issues/2320 \].

**Comment by vollkorntomate** [comment]
Am I wrong or do the Rust SDK and Element X already *almost* support multiple accounts?
I had a quick naive look into the code and from my understanding, there is nothing in the way of simply creating a second (third, ...) user session through the app, and all the storage etc. is already taken care of.

The only thing that doesn't seem to be possible yet is to restore a session for a given userID, but that looks to me like a rather small change in the `UserSessionStoreProtocol` and its implementation. And of course, the UIs for the account management are still missing, but that should be quite straightforward.

Another issue that I can think of is notification management, but I'd view that as a second step once we are able to add multiple accounts.

**Comment by ara4n** [comment]
you are right. it "just" needs the auth, login, registration, acct management, notif, account switching UI, etc. we are also desperate to have it as a feature tho; it's been excruciating not having it in classic Element.

**Reply by claell** [reply]
So why not put actions to your statements? If you want this, why was this converted to a discussion instead of an issue which could act as a meta issue for the next steps? The things you listed as requirements can be linked to that meta issue, then.

Would you (honestly curious) mind explaining, how you (at vector-im, since you use "we") can at the same time be desperate to have it as a feature while moving issues to track progress and plan actionable paths to discussions?

**Reply by awtj8o81ryywg793** [reply]
I understand that your reply was written _some_ time ago, but I'll take a crack at answering you.

> Why was this converted to a discussion instead of an issue that could act as a meta issue for the next steps?

Discussions provide a platform for addressing new features or major cross-platform changes, while issues in specific repositories typically focus on bugs, enhancements to already existing functionality, or potential security vulnerabilities.

If you open an issue in the [element-hq/element-x-android](https://github.com/element-hq/element-x-android) repository and select the "Enhancement request" option, you'll be greeted with:

> Thank you for taking the time to propose an enhancement to an existing feature. If you would like to propose a new feature or a major cross-platform change, please [start a discussion here](https://github.com/element-hq/element-meta/discussions/new?category=ideas).

Similarly, if you open an issue in the [element-hq/element-android](https://github.com/element-hq/element-android) repository and click the "Enhancement or feature request" option, you'll be redirected to start a discussion as well.

> The things you listed as requirements can be linked to that meta issue, then.

See above.

> Would you (honestly curious) mind explaining how you (at Vector.im, since you use "we") can be desperate to have it as a feature while moving issues to track progress and plan actionable paths to discussions?

You'll be happy to know that Element X, at least on Android, has experimental support for multiple accounts!

You may find [this](https://github.com/element-hq/element-x-android/issues/190#issuecomment-1487044459) comment by @manuroe to be of interest.

Note that I'm not completely certain the flows mentioned previously were present when you posted your reply, but given that this discussion was converted from an issue, I'm fairly sure that was the case.

**Reply by claell** [reply]
Thanks for taking the time to reply.

However, your answer kind of misses the main point I wanted to make with my questions. My question was basically: If one is that desperate to have it as a feature, why move it to backlog, instead of creating actionable items for the many steps required to reach it?

**Reply by awtj8o81ryywg793** [reply]
Quoted from the [above](https://github.com/element-hq/element-meta/discussions/1832#discussioncomment-14685659) comment:
> If one is that desperate to have it as a feature, why move it to backlog, instead of creating actionable items for the many steps required to reach it?

I have no additional insight into Element matters beyond what is publicly available. However, in the linked comment in my [previous reply](https://github.com/element-hq/element-meta/discussions/1832#discussioncomment-14682124), it was mentioned that the necessary work to ensure multiple account functionality is operational and stable was already underway before the reply we're discussing was posted.

For what it's worth, I wouldn't call discussions a "backlog." As for why, please see my [last](https://github.com/element-hq/element-meta/discussions/1832#discussioncomment-14682124) reply.

Anyway, support for multiple accounts is starting to land in Element X, so 🤷‍♂️.

**Comment by darkdragon-001** [comment]
> In fact, GNOME's [Fractal](https://gitlab.gnome.org/GNOME/fractal) matrix client makes use of the Rust SDK and already supports multi-account! https://gitlab.gnome.org/GNOME/fractal/-/tree/main/src/account_switcher

~ https://github.com/vector-im/element-x-android/issues/190#issuecomment-1466174356

**Reply by kevincox** [reply]
This is very cool but honestly not what I am looking for in multi-account. Which highlights that there are different possible UX and making the right decision could be difficult.

I want to see the different accounts being fully integrated. I want all of my chats to be visible in one list, I have one "new message" button that can message from either account. Having a top level account switcher is kind of the minimum possible. You can log into both and get notifications at the same time. But other than that they are completely separate.

It is kind of ironic that with bridging I can have multiple IRC accounts, multiple Facebook accounts and whatever else in one client integrated. But one other Matrix account and it is almost completely separate. Maybe I just need a Matrix-to-Matrix bridge.

**Reply by mpeter50** [reply]
I think such should be optional, because I think for most of us that would be a mess. Also, I dont want to see work messages when not at work, and also when I'm recording or sharing my screen (even on Android, yes) I dont want my personal chats to be visible.

Also, having all accounts in one view would mean that they need to find a way to put an account selection button to many different places, 1) to make it clear which of your accounts will take an action, and 2) to make it selectable which of your accounts will take an action.

Is there a messaging service that works the way you describe? Checking how does it exactly work could clear up some things.

**Reply by Porkepix** [reply]
> Is there a messaging service that works the way you describe? Checking how does it exactly work could clear up some things.

Several XMPP clients such as Gajim, Dino or Conversations. And for many actions it's natural: you speak from the account you joined, and for 1:1 chat they're contacts to one account. You mostly have to select when you add a contact or join a chan.

**Reply by kevincox** [reply]
> Is there a messaging service that works the way you describe?

Yeah, Matrix does when I bridge all of my accounts into there :) It is great.

But I think very few messaging apps support the concept of multiple accounts. This is probably because it is often quite complicated. That is why for many use cases it makes sense to avoid adding a new top-level item to the hierarchy.

> Also, I dont want to see work messages when not at work

Yeah, this is also an important UX feature. I can definitely see the benefit of having groups of accounts that you can hide and show (possibly together). In fact this is a very similar experience to a lot of things you may want to do with spaces as well. I can also imagine having multiple work accounts that I want to hide and show together.

So I think this is a useful feature, but I don't think tying it 1:1 to accounts is the right away to go about it.

> Also, having all accounts in one view would mean that they need to find a way to put an account selection button to many different places

Yeah, I thought about this a bit but I actually think it is pretty few. What I'm imagining is that each room belongs to an account (I suspect if you are in the same room in two accounts it shows up separately, but other UX options could make sense too). So you only really need to select an account when creating a new room. You may want to look at beeper.com here. When you create a new chat you select which network you want to create it on. This is more or less the same concept as deciding which account to use when creating the room.

Very few things in Matrix take place outside of rooms, so I don't think you explicitly need to select that often. For things like changing your display name and picture yes, but this is obviously account specific. Notification settings you would likely want to set per-account as well (although I can imagine some interface to unify them into one global setting which gets fanned out).

**Comment by DraconicNEO** [comment]


**Reply by mobergmann** [reply]
The App is still in its early stages, as you can read at the top of the README:
> This app is currently in a pre-alpha release stage with only basic functionalities.

Therefore there are a lot of features still "missing".

**Reply by DraconicNEO** [reply]
Good point, I wrote that out of frustration but I probably shouldn't have since the app is in early stages, I have deleted my rather rude comment

**Comment by Mek101** [comment]
Any ETA? May 2047?

**Reply by Iron-Squid** [reply]
It definitely has been a while.

**Reply by awtj8o81ryywg793** [reply]
See [this](https://github.com/element-hq/element-meta/discussions/1832#discussioncomment-14682145) comment.

**Comment by xttlegendapi** [comment]
Pls

**Reply by awtj8o81ryywg793** [reply]
See [this](https://github.com/element-hq/element-meta/discussions/1832#discussioncomment-14682145) comment.

**Comment by dawagner** [comment]
Any news or alternative? I have one personal account and one at $DAYJOB and would like to access both on my android device.

**Reply by dawagner** [reply]
Re:alternatives : I turned to the "Shelter" app (available on FDroid), which creates a so-called "work profile" in which you can clone apps. But it's under limited maintenance and has some drawbacks (e.g. you can't open links out if the box).

**Reply by KaKi87** [reply]
> But it's under limited maintenance

Which is not an issue;

> and has some drawbacks (e.g. you can't open links out if the box).

Which is annoying, in addition to not sharing files, clipboard, and any other app configuration, like the keyboard.

Same with the official experimental [app cloning feature introduced in Android 14](https://www.xda-developers.com/android-14-app-cloning/) : it uses profiles as well under the hood.

**Reply by The-Nickname** [reply]
Re: Alternatives:
You could use [FluffyChat](https://matrix.org/ecosystem/clients/fluffychat) or [Tammy](https://matrix.org/ecosystem/clients/tammy) they support Multiple Accounts, or any other App that support that, see: https://matrix.org/ecosystem/clients/ (Filter by Features and Plattform etc)

**Reply by florian-bellencontre** [reply]
> Re: Alternatives: You could use [FluffyChat](https://matrix.org/ecosystem/clients/fluffychat) or [Tammy](https://matrix.org/ecosystem/clients/tammy) they support Multiple Accounts, or any other App that support that, see: https://matrix.org/ecosystem/clients/ (Filter by Features and Plattform etc)

Not available on french apple store :(

**Reply by The-Nickname** [reply]
> Not available on french apple store :(

You sure about that? https://apps.apple.com/app/fluffychat/id1551469600?l=fr-FR otherwise maybe you can check why they arent available in France, the Google Play version doesnt seem restricted either so Idk.

**Reply by awtj8o81ryywg793** [reply]
See [this](https://github.com/element-hq/element-meta/discussions/1832#discussioncomment-14682145) comment.

**Comment by KoljaFrahm** [comment]
I understand the first reply as it being almost ready? So is someone working on it?

**Reply by awtj8o81ryywg793** [reply]
See [this](https://github.com/element-hq/element-meta/discussions/1832#discussioncomment-14682145) comment.

**Comment by awtj8o81ryywg793** [comment]
Element X now has experimental support for multiple accounts.

Edit: It seems that only Element X for Android has such support.

**Reply by florian-bellencontre** [reply]
Where did you see that? On iOS or Android?

I don't see anything on iOS (App Store or TestFlight version).

**Reply by bricewge** [reply]
I think it's related to this PR https://github.com/element-hq/element-x-android/pull/5285. And has been announced on the latest release notes: https://github.com/element-hq/element-x-android/releases/tag/v25.10.0

But I wasn't able to try it on Android, with v25.10.0, it looks like it's behind a build flag.

**Reply by tobiasBora** [reply]
What is the progress on this 1 year later?

**Reply by awtj8o81ryywg793** [reply]
Quoted from a [previous comment](https://github.com/element-hq/element-meta/discussions/1832#discussioncomment-15746516):

> What is the progress on this 1 year later?

I cannot speak for Element X on iOS, but on Android, you can enable multi-account support within the developer options (as it is still considered experimental).

**Reply by tobiasBora** [reply]
Oh cool, so I don't need to compile it with a specific flag? I'll give it a try, thanks!

**Reply by janvarga** [reply]
I have Element X installed on my Android 9 phone (installed via Play Store) but I can't find any developer options.

**Reply by mobergmann** [reply]
In the settings menu you need to press the version number (at the bottom of the menu) repeatedly to unlock the developer options settings.

**Reply by janvarga** [reply]
Oh wow, thanks!

**Reply by Kagee** [reply]
Is there any way i can contribute to making this feature not "EXPERINENTAL and UNSTABLE"? I need this feature for my normal Matrix account, and one, soon two non-federated work accounts.

**Reply by awtj8o81ryywg793** [reply]
Quoted from a [previous comment](https://github.com/element-hq/element-meta/discussions/1832#discussioncomment-17446873):

> Is there any way i can contribute to making this feature not "EXPERINENTAL and UNSTABLE"? I need this feature for my normal Matrix account, and one, soon two non-federated work accounts.

I’d recommend looking at the issue tracker, whether that’s for Element X for Android or Element X for iOS, and diving in with a helping hand if there are any outstanding points. It probably wouldn’t be a bad idea to join each project’s respective rooms and ask around too. Reporting, and possibly fixing, any bugs you come across would also be a great help.

**Comment by alfredonodo** [comment]
Any news for this fundamental feature?
Thank you

**Reply by wrenix** [reply]
Read https://github.com/element-hq/element-meta/discussions/1832#discussioncomment-14682145
